articles/contributing/api-design-principles.md

```diff
@@ -0,0 +1,426 @@
+      example:
+    - When renaming an operation called `EstimateExpectation` to
+        `EstimateAverage`, introduce a new operation called
+        `EstimateExpectation` that calls the original operation at
+        its new name, so that existing code can continue to work
+        correctly.
+  - **DO** use the @"microsoft.quantum.core.deprecated" attribute to communicate
+      deprecations to the user.
+  - When renaming an operation or function, **DO** provide the new
+      name as a string input to `@Deprecated`.
+  - **DON'T** remove existing functions or operations without a
+      deprecation period of at least six months for preview releases,
+      or at least two years for supported releases.
+
+- **DO** follow the common [Q\# style guide](xref:microsoft.quantum.contributing.style)
+    whenever applicable.
+
+## Functions and Operations
+
+- **DO** ensure that every function and operation has a single
+    well-defined purpose within the API.
+  - **DON'T** expose functions and operations that perform multiple
+      unrelated tasks.
+
+- **DO** design functions and operations to compose well with other
+    functions and operations, both in the same API and in previously
+    existing libraries.
+
+- **DO** design functions and operations to be as reusable as
+    possible, and to anticipate future needs.
+
+  - **DON'T** expose purely deterministic classical logic as an
+      operation; **DO** expose purely deterministic classical logic as
+      a function. For example:
+
+    - A subroutine which squares its floating-point input can be
+        written deterministically, and so should be exposed to the
+        user as `Squared : Double -> Double` rather than as an
+        operation `Square : Double => Double`. This allows for the
+        subroutine to be called in more places (e.g.: inside of
+        other functions), and provides useful optimization
+        information to the compiler that can affect performance and
+        optimizations.
+
+    - `ForEach<'TInput, 'TOutput>('TInput => 'TOutput, 'TInput[]) => 'TOutput[]`
+        and `Mapped<'TInput, 'TOutput>('TInput -> 'TOutput, 'TInput[]) -> 'TOutput[]`
+        differ in the guarantees made with respect to
+        determinism; both are useful in different circumstances.
+
+    - API routines that transform the application of quantum
+        operations can often be carried out in a deterministic
+        fashion and hence can be made available as functions such as
+        `CControlled<'T>(op : 'T => Unit) => ((Bool, 'T) => Unit)`.
+
+    - **DO** generalize the input type as much as reasonable for each
+        function and operation, using type parameters as needed. For
+        example:
+      - `ApplyToEach` has type `<'T>(('T => Unit), 'T[]) => Unit`
+          rather than the specific type of its most common
+          application, `((Qubit => Unit), Qubit[]) => Unit`.
+
+- **DO** choose input and output types for functions and operations
```

Comment:
Same, indentation is probably not the best option to group related principles. Maybe create a sub-section?

Reply:
Refactored significantly, I think that should help here. Please feel free to reopen if you feel there's more work to be done for this one. Thanks!